beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandler.java

```diff
@@ -312,19 +322,40 @@ public SafeFuture<Optional<BlockContainer>> createUnsignedBlock(
     if (isSyncActive()) {
       return NodeSyncingException.failedFuture();
     }
+    final BlockProductionPerformance blockProductionPerformance =
+        blockProductionPerformanceTrackingEnabled
+            ? new BlockProductionPerformanceImpl(timeProvider, slot)
+            : BlockProductionPerformance.NOOP;
     return forkChoiceTrigger
-        .prepareForBlockProduction(slot)
+        .prepareForBlockProduction(slot, blockProductionPerformance)
         .thenCompose(__ -> combinedChainDataClient.getStateAtSlotExact(slot))
+        .thenPeek(
+            maybeState -> {
+              maybeState.ifPresent(
+                  state ->
+                      blockProductionPerformance.slotTime(
+                          () -> secondsToMillis(spec.computeTimeAtSlot(state, slot))));
+              blockProductionPerformance.getStateAtSlot();
+            })
         .thenCompose(
-            blockSlotState -> createBlock(slot, randaoReveal, graffiti, blinded, blockSlotState));
+            blockSlotState ->
+                createBlock(
+                    slot,
+                    randaoReveal,
+                    graffiti,
+                    blinded,
+                    blockSlotState,
+                    blockProductionPerformance))
+        .thenPeek(__ -> blockProductionPerformance.complete());
```

Comment:
Maybe it makes sense to put it in something like `alwaysRun`, so we could still recover something in timeout something cases?

Reply:
ye makes sense